vsintegration/tests/FSharp.Editor.Tests/Refactors/AddExplicitReturnType.fs

```diff
@@ -0,0 +1,180 @@
+﻿module FSharp.Editor.Tests.Refactors.AddExplicitReturnType
+
+open Microsoft.VisualStudio.FSharp.Editor
+open Xunit
+open System
+open System.Collections.Immutable
+open System.Text.RegularExpressions
+
+open Microsoft.CodeAnalysis
+open Microsoft.CodeAnalysis.CodeFixes
+open Microsoft.CodeAnalysis.Text
+open Microsoft.VisualStudio.FSharp.Editor
+open Microsoft.VisualStudio.FSharp.Editor.CancellableTasks
+
+open FSharp.Compiler.Diagnostics
+open FSharp.Editor.Tests.Helpers
+
+open Microsoft.CodeAnalysis.CodeRefactorings
+open NUnit.Framework
+open Microsoft.CodeAnalysis.CodeActions
+open System.Collections.Generic
+open Microsoft.VisualStudio.LanguageServices
+open FSharp.Editor.Tests.Refactors.RefactorTestFramework
+open Microsoft.Build.Utilities
+open System.Threading
+open FSharp.Test.ReflectionHelper
+open Microsoft.Build.Utilities
+open FSharp.Test.ProjectGeneration.ProjectOperations
+open FSharp.Compiler.Symbols
+open Xunit
+open System.Runtime.InteropServices
+
+[<Theory>]
+[<InlineData(":int")>]
+[<InlineData(" :int")>]
+[<InlineData(" : int")>]
+[<InlineData(" :    int")>]
+let ``Refactor should not trigger`` (shouldNotTrigger: string) =
+    task {
+        let symbolName = "sum"
+
+        let code =
+            $"""
+            let sum a b {shouldNotTrigger}= a + b
+            """
+
+        use context = TestContext.CreateWithCode code
+
+        let spanStart = code.IndexOf symbolName
+
+        let! actions = tryGetRefactoringActions code spanStart context (new AddExplicitReturnType())
+
+        do Assert.Empty(actions)
+    }
+
+[<Fact>]
+let ``Correctly infer int as explicit return type`` () =
+    task {
+
+        let symbolName = "sum"
+
+        let code =
+            """
+            let sum a b = a + b
+            """
+
+        use context = TestContext.CreateWithCode code
+
+        let spanStart = code.IndexOf symbolName
+
+        let! newDoc = tryRefactor code spanStart context (new AddExplicitReturnType())
+
+        do! AssertHasSpecificExplicitReturnType symbolName "int" newDoc context.CT
+    }
+
+[<Theory>]
+[<InlineData("(a:float) (b:int)", "float")>]
+[<InlineData("a:int b:int", "int")>]
+let ``Infer explicit return type`` (functionHeader: string) (returnType: string) =
+    task {
+
+        let symbolName = "sum"
+
+        let code =
+            $"""
+            let sum {functionHeader}= a + b
+            """
+
+        use context = TestContext.CreateWithCode code
+
+        let spanStart = code.IndexOf(symbolName)
+
+        let! newDoc = tryRefactor code spanStart context (new AddExplicitReturnType())
+        do Assert.NotNull(code)
```

Comment:
btw. I am quite good at stumbling upon strange behaviours apparently XD @psfinaki  Quiz. Why is spanStart=0 at line 91 ? And why do I have to assert the value of code to even get a value ? 
Case 1 with Line 94 commented out : 
![image](https://github.com/dotnet/fsharp/assets/19798013/4ffb4d61-dd5c-422f-8589-3952862c027a)
Case 2 with Line 94 not commented out: 
![image](https://github.com/dotnet/fsharp/assets/19798013/f4c946fa-fcf8-4d86-8974-ee8b7d5000a2)

You can see an evaluation of the expression in the immediate window down right. Without the assert, code is null. That should  absolutely never happen since code is used. But even if it is not null, spanStart gives 0 in the code but 18 in the immediate window ?

Reply:
Hmm that said sounds like some issues with async processing or lazy eval. Maybe yeah, try making tests sync and seeing if this still reproduces.

Reply:
Yeah, seems to be an async problem, I made unit tests to reproduce it. But any kind of using the values and it works. For example, breaking before the task binding expression:
![image](https://github.com/dotnet/fsharp/assets/19798013/b23da289-1d80-4f85-8a6d-ca0730115adc)
And after:
![image](https://github.com/dotnet/fsharp/assets/19798013/99c57794-a910-434c-abee-281a94ab04a3)

No problem on the non computation expression version.
before:
![image](https://github.com/dotnet/fsharp/assets/19798013/f0952d22-e498-4146-9276-4c52610d7c03)
After waiting for the result : 
![image](https://github.com/dotnet/fsharp/assets/19798013/4da94314-a225-4361-b9bc-bf1a6fbc699f)

Reply:
Yeah so I am glad it helped. F# is very lazy which is fine for actual UI scenarios but in tests we sometimes force sequences and values when we needed.

Reply:
@psfinaki Is this worth opening an issue ? Seems at least very counter intuitive

Reply:
Well if you manage to isolate some particularly weird behavior, sure. Although I wouldn't expect anything happening against the spec here.

Reply:
@psfinaki I would say that showing the wrong value in the debugger should be against the spec XD